### PR TITLE
Add caching.

### PR DESCRIPTION
**Pull request by lachlansneff**
This pr adds the ability to compile wasm into a `Cache`, and then you can store cache to a file, load it from a file, and convert it to a `Module`.

### CODE REVIEW

File: lib/runtime-core/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    pub fn info(&self) -> &ModuleInfo {
+        &self.inner.info
+    }
+
+    pub fn wasm_hash(&self) -> &[u8; 32] {
+        &self.wasm_hash
+    }
+
+    #[doc(hidden)]
+    pub fn consume(self) -> (ModuleInfo, Vec<u8>, Memory) {
```
Comment (syrusakbary) [inline lib/runtime-core/src/cache.rs]: What this function is for?
Can we make it public only for the crate?

Comment (lachlansneff) [reply lib/runtime-core/src/cache.rs]: Not really, since it's used in the clif backend.

File: lib/runtime-core/src/backing.rs [hunk 2]
```diff
@@ -170,8 +171,12 @@ impl LocalBacking {
...
                     table.anyfunc_direct_access_mut(|elements| {
                         for (i, &func_index) in init.elements.iter().enumerate() {
-                            let sig_index = module.func_assoc[func_index];
-                            let sig_id = vm::SigId(sig_index.index() as u32);
+                            let sig_index = module.info.func_assoc[func_index];
+                            let signature = &module.info.signatures[sig_index];
+                            let sig_id = vm::SigId(
+                                SigRegistry.lookup_sig_index(Arc::clone(&signature)).index() as u32,
+                            );
+                            println!("func_index: {:?}, sig_id: {:?}", func_index, sig_id);
```
Comment (syrusakbary) [inline lib/runtime-core/src/backing.rs]: `debug!` instead?

File: lib/runtime-core/src/backing.rs [hunk 3]
```diff
@@ -205,8 +210,12 @@ impl LocalBacking {
...
```
Comment (syrusakbary) [inline lib/runtime-core/src/backing.rs]: `debug!` ?

File: lib/runtime-core/src/sys/unix/memory.rs [hunk 4]
```diff
@@ -11,14 +11,94 @@ unsafe impl Sync for Memory {}
...
+                protection.to_protect_const() as i32,
+                libc::MAP_PRIVATE,
+                raw_fd.0,
+                0,
+            )
+        };
+
+        let elapsed = start.elapsed();
+
+        println!("mmap time: {:?}", elapsed);
```
Comment (syrusakbary) [inline lib/runtime-core/src/sys/unix/memory.rs]: `debug!`

File: lib/spectests/examples/simple/main.rs [hunk 5]
```diff
@@ -13,8 +15,38 @@ use wasmer_runtime_core::{
 static EXAMPLE_WASM: &'static [u8] = include_bytes!("simple.wasm");
 
 fn main() -> Result<()> {
+    let compiler = CraneliftCompiler::new();
     let wasm_binary = wat2wasm(IMPORT_MODULE.as_bytes()).expect("WAST not valid or malformed");
-    let inner_module = wasmer_runtime_core::compile_with(&wasm_binary, &CraneliftCompiler::new())?;
+
+    let inner_module = {
+        println!("Creating cache...");
```
Comment (syrusakbary) [inline lib/spectests/examples/simple/main.rs]: `debug!`

File: lib/spectests/examples/simple/main.rs [hunk 6]
```diff
@@ -13,8 +15,38 @@ use wasmer_runtime_core::{
...
+
+        let cache = wasmer_runtime_core::compile_to_cache_with(&wasm_binary, &compiler)?;
+
+        println!("Writing cache to disk...");
```
Comment (syrusakbary) [inline lib/spectests/examples/simple/main.rs]: `debug!`

File: lib/spectests/examples/simple/main.rs [hunk 7]
```diff
@@ -13,8 +15,38 @@ use wasmer_runtime_core::{
...
+        cache.store("import_module.wasmer").unwrap();
+
+        println!("Opening cache on disk...");
```
Comment (syrusakbary) [inline lib/spectests/examples/simple/main.rs]: `debug!`

File: lib/spectests/examples/simple/main.rs [hunk 8]
```diff
@@ -13,8 +15,38 @@ use wasmer_runtime_core::{
...
+
+        let start = Instant::now();
+
+        // println!("Loading cache...");
+        let inner_module =
+            unsafe { wasmer_runtime_core::load_cache_with(cache, &compiler).unwrap() };
+
+        let elapsed = start.elapsed();
+
+        println!("time to load cache: {:?}", elapsed);
```
Comment (syrusakbary) [inline lib/spectests/examples/simple/main.rs]: `debug!`

File: lib/runtime/src/lib.rs [hunk 9]
```diff
@@ -100,12 +98,26 @@ pub mod wasm {
...
     pub use wasmer_runtime_core::units::{Bytes, Pages};
 }
 
+#[cfg(feature = "default-compiler")]
+mod cache;
+
+#[cfg(feature = "default-compiler")]
+use wasmer_runtime_core::backend::Compiler;
+
+#[cfg(feature = "default-compiler")]
```
Comment (syrusakbary) [inline lib/runtime/src/lib.rs]: Shouldn't this be `cache` ?

File: lib/runtime/src/lib.rs [hunk 10]
```diff
@@ -100,12 +98,26 @@ pub mod wasm {
...
+    pub use wasmer_runtime_core::error::*;
+}
+
 pub mod units {
     //! Various unit types.
-
     pub use wasmer_runtime_core::units::{Bytes, Pages};
 }
 
+#[cfg(feature = "default-compiler")]
```
Comment (syrusakbary) [inline lib/runtime/src/lib.rs]: Shouldn't this be `cache` ?

File: lib/runtime/src/lib.rs [hunk 11]
```diff
@@ -143,11 +154,47 @@ pub fn compile(wasm: &[u8]) -> error::CompileResult<Module> {
...
 /// `error::RuntimeError` (all combined into an `error::Error`),
 /// depending on the cause of the failure.
-#[cfg(feature = "wasmer-clif-backend")]
+#[cfg(feature = "default-compiler")]
 pub fn instantiate(wasm: &[u8], import_object: &ImportObject) -> error::Result<Instance> {
     let module = compile(wasm)?;
     module.instantiate(import_object)
 }
 
+/// Compile wasm into a [`Cache`] that can be stored to a file or
```
Comment (syrusakbary) [inline lib/runtime/src/lib.rs]: Maybe we should only export this if cache is enabled?

File: lib/runtime-core/src/backing.rs (hunk 2 above)
Comment (lachlansneff) [reply lib/runtime-core/src/backing.rs]: Ah, yep, forgot to remove this print.

File: lib/runtime-core/src/backing.rs (hunk 3 above)
Comment (lachlansneff) [reply lib/runtime-core/src/backing.rs]: ditto

File: lib/runtime-core/src/sys/unix/memory.rs (hunk 4 above)
Comment (lachlansneff) [reply lib/runtime-core/src/sys/unix/memory.rs]: ditto again

File: lib/spectests/examples/simple/main.rs (hunk 5 above)
Comment (lachlansneff) [reply lib/spectests/examples/simple/main.rs]: and again

File: lib/runtime/src/lib.rs (hunk 9 above)
Comment (lachlansneff) [reply lib/runtime/src/lib.rs]: cache relies on there being a default-compiler.

File: lib/runtime/src/lib.rs (hunk 10 above)
Comment (lachlansneff) [reply lib/runtime/src/lib.rs]: see above

File: lib/runtime/src/lib.rs (hunk 11 above)
Comment (lachlansneff) [reply lib/runtime/src/lib.rs]: I swear I did feature-gate this, will fix.

**Pull request merged**